### PR TITLE
Handle relative git-url-paths correctly when --directory

### PR DESCRIPTION
**Pull request by sigurdm**

### CODE REVIEW

**Review by jonasfj — APPROVED** [review]

File: lib/src/source/git.dart [hunk 1]
```diff
@@ -249,34 +283,43 @@ class BoundGitSource extends CachedSource {
...
           .run(['show', '$revision:$pubspecPath'], workingDir: repoPath);
     } on git.GitException catch (_) {
       fail('Could not find a file named "$pubspecPath" in '
-          '${ref.description['url']} $revision.');
+          '$actualPath $revision.');
     }
 
     return Pubspec.parse(lines.join('\n'), systemCache.sources,
-        expectedName: ref.name);
+        expectedName: ref.name, location: Uri.parse(actualPath));
```
Comment (jonasfj) [inline lib/src/source/git.dart:322]: Let's not add this: `, location: Uri.parse(actualPath)`

File: lib/src/command/add.dart [hunk 2]
```diff
@@ -322,7 +322,7 @@ class AddCommand extends PubCommand {
       }
 
       packageRange = cache.sources['git']
-          .parseRef(packageName, git)
+          .parseRef(packageName, git, containingPath: entrypoint.pubspecPath)
           .withConstraint(constraint ?? VersionConstraint.any);
       pubspecInformation = {'git': git};
```
Comment (jonasfj) [inline lib/src/command/add.dart:327]: `git` contains `'url'` which must be relative to `entrypoint.pubspecPath`.

Comment (jonasfj) [reply lib/src/command/add.dart:327]: We have the same problem in `path`.

**Review by sigurdm — COMMENTED** [review]
Could you PTAL - especially the add-part?

File: lib/src/global_packages.dart [hunk 3]
```diff
@@ -90,9 +90,14 @@ class GlobalPackages {
     // dependencies. Their executables shouldn't be cached, and there should
     // be a mechanism for redoing dependency resolution if a path pubspec has
     // changed (see also issue 20499).
+    PackageRef ref;
+    try {
+      ref = cache.git.source.parseRef(name, {'url': repo}, containingPath: '.');
```
Comment (jonasfj) [inline lib/src/global_packages.dart:95]: Technically, we're going to write the `pubspec.yaml` somewhere other than `.`, so the `containingPath` should be something like `$PUB_CACHE/global_packages/<package>/`, shouldn't it?

I'm not sure it matters here?

**Review by jonasfj — APPROVED** [review]

File: lib/src/global_packages.dart (hunk 3 above)
Comment (sigurdm) [reply lib/src/global_packages.dart:95]: We don't write it to disk (yet), so it should not matter...

**Pull request merged**